fpsd/features.py

```diff
@@ -0,0 +1,768 @@
+        by the position-based feature generation functions.
+
+        Args:
+            outgoing_only [boolean]: describes whether or not we should
+                only use outgoing cells to create this table
+        """
+
+        self.drop_table("packet_positions")
+
+        if outgoing_only:
+            where_only_outgoing = "WHERE ingoing <= false"
+        else:
+            where_only_outgoing = ""
+
+        query = ("CREATE TEMP TABLE packet_positions AS             "
+                 "(SELECT exampleid, position, ingoing              "
+                 "  FROM (                                          "
+                 "    SELECT                                        "
+                 "      ROW_NUMBER() OVER                           "
+                 "      (PARTITION BY exampleid ORDER BY t_trace)   "
+                 "      AS position,                                "
+                 "      t.*                                         "
+                 "    FROM raw.frontpage_traces t)                  "
+                 "x {} );").format(where_only_outgoing)
+
+        self.engine.execute(query)
+
+    def _create_table_outgoing_cell_positions(self, num_cells=500):
+        """This method takes the first num_cells rows for each exampleid
+        in packet_positions and creates a temporary table
+        first_(num_cells)_outgoing_packet_positions with the following
+        format:
+
+        exampleid  | position |  outgoing_cell_order
+        (integer)  | (bigint) |  (bigint)
+        -----------------------------------------------
+        9          | 4        | 1
+        9          | 5        | 2
+        9          | 15       | 3
+
+        For example, the third outgoing cell was at position 15 in the
+        trace. This table is used by the position-based feature
+        generation functions.
+        """
+        table_name = "first_{}_outgoing_packet_positions".format(num_cells)
+        self.drop_table(table_name)
+
+        query = ("CREATE TEMP TABLE first_{n}_outgoing_packet_positions AS "
+                 "(SELECT exampleid, outgoing_cell_order, position      "
+                 "  FROM (                                                 "
+                 "    SELECT                                               "
+                 "      ROW_NUMBER() OVER                                  "
+                 "      (PARTITION BY exampleid ORDER BY position)         "
+                 "      AS outgoing_cell_order,                         "
+                 "      t.*                                                "
+                 "    FROM packet_positions t) x                           "
+                 "  WHERE x.outgoing_cell_order <= {n}                  "
+                 ");").format(n=num_cells)
+
+        self.engine.execute(query)
+
+    def get_exampleids(self):
+        """Get list of exampleids"""
+        query = "SELECT DISTINCT exampleid FROM raw.frontpage_traces"
```

Comment:
The `DISTINCT` is necessary here: `exampleid` is the primary key in `frontpage_examples`, `traceid is the primary key in `frontpage_traces`